File: day10/2_K_Diff_Pairs.java
/*
Auther - Saurabh Verma
Problem link - https://leetcode.com/problems/k-diff-pairs-in-an-array/
Approach - use hashmap to store frequency of each value to handle k ==0 case
*/
class Solution {
    public int findPairs(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>();
        int count = 0;
        for(int i: nums)
            map.put(i, map.getOrDefault(i, 0)+1);
        for(Map.Entry<Integer, Integer> e: map.entrySet()){
            
            if(k == 0 && e.getValue() > 1)
                count++;
            else if(k>0 && map.containsKey(e.getKey() + k))
                count++;
        }
        return count;
    }
}

File: day10/2_Kdiff_pairs.py
# created by KUMAR SHANU

# 2. K-diff Pairs in an Array
# https://leetcode.com/problems/k-diff-pairs-in-an-array/
"""Store frequency in dictionary"""
from collections import defaultdict


class Solution:
    def findPairs(self, nums: List[int], k: int) -> int:
        count = 0
        d = defaultdict(int)
        for num in nums:
            d[num] += 1

        for i in d:
            if k == 0 and d[i] > 1:
                count += 1
            elif k > 0 and ((i + k) in d):
                count += 1
        #print(d)
        return count
